### DISCUSSION TITLE
Where is the schema for fastify.route's options.schema.{body, querystring, params, response} stored in?

### DISCUSSION THREAD

**Discussion by billxinli** [fastify/fastify#4384] [Q&A]
Hello! 

I am attempting to upgrade a few of our APIs built with Fastify v3 to Fastify v4. However, I am having a slight issue with how the schemas for a route's `body`, `querystring`, `params`, and `responses` are being validated.

We are taking an OpenAPI v3.0.3 spec and breaking it down into the proper parts for the body, querystring, params, and responses and automatically building a set of routes in fastify v3. So in essence, we provide a spec, with a [plugin](https://github.com/ahmadnassri/node-oas-fastify/blob/master/lib/plugin.js#L91), and all the routes are setup for us.

An example `fastify.route` call would look like:

```
{
  "method": "PATCH",
  "url": "/entities/:entity_id",
  "schema": {
    "params": {
      "type": "object",
      "required": [
        "entity_id"
      ],
      "properties": {
        "entity_id": {
          "type": "string",
          "format": "uuid"
        }
      }
    },
    "query": {
      "type": "object",
      "required": [],
      "properties": {}
    },
    "headers": {
      "type": "object",
      "required": [
        "content-type"
      ],
      "properties": {
        "content-type": {
          "const": "application/json"
        }
      }
    },
    "response": {
      "201": {
        "type": "object",
        "properties": {
          "entity_id": {
            "type": "string",
            "format": "uuid",
            "description": "the entity id",
            "example": "75bba958-527d-494e-91f8-1f6fdeccf45b",
            "readOnly": true
          }
        }
      }
    },
    "body": {
      "type": "object",
      "required": [
        "entity_import_url"
      ],
      "properties": {
        "entity_import_url": {
          "type": "string",
          "description": "the entity import url",
          "example": "https://www.nasa.gov/sites/default/files/thumbnails/image/main_image_star-forming_region_carina_nircam_final-5mb.jpg"
        }
      }
    }
  }
}
```

However, since the spec came from OpenAPI spec, we have additional attributes here, specifically the `example` fields. 

Upon upgrading to Fastify v4, we are seeing that it fails the schema validation.

If we set the custom AJV option for `strictSchema` to `log` like below, 

```
ajv: {
  customOptions: {
    allErrors: true,
    strictSchema: 'log' //example
  }
}
```

We can see the error `strict mode: unknown keyword: "example"`.

My question, other than setting `strictSchema` to `log` or false altogether. Is it possible to allow these additional attributes that are also valid from OpenAPI spec?

Thank you!

**Comment by Eomm** [comment]
> Upon upgrading to Fastify v4, we are seeing that it fails the schema validation.

It is due the AJV version from v6 to v8

> My question, other than setting strictSchema to log or false altogether. Is it possible to allow these additional attributes that are also valid from OpenAPI spec?

No, AJV supports some OpenAPI keywords. Ref:

- https://ajv.js.org/json-schema.html#openapi-support
- https://ajv.js.org/json-schema.html#metadata-keywords

Unfortunately, JSON Schema and OpenAPI are different standards that are merging later than sooner

**Reply by billxinli** [reply]
So it seems like the best way forward is to strip out any OpenAPI spec keywords that are not supported.

The alternative is to turn off strict validation. Which may have other effects?
